### PR TITLE
BI-1183 - Import Validations: Ontology-specific observation criteria

### PR DESCRIPTION
**Pull request by nickpalladino**
# Description
**Story:** [BI-1183](https://breedinginsight.atlassian.net/browse/BI-1183?atlOrigin=eyJpIjoiMDI5YmU2YjI5ZjIwNDJhMGI0ZjIyN2FkMzhlMmNhYzYiLCJwIjoiaiJ9)

- Added import template dynamic column processing for phenotypes
- Added validations of phenotype column data

# Dependencies
- bi-web BI-1183

# Testing
- Example test file included. Test validations to ensure they are working according to card acceptance criteria.
- [demo.xlsx](https://github.com/Breeding-Insight/bi-api/files/9709430/demo.xlsx)

# Checklist:
- [x] I have performed a self-review of my own code
- [x] I have tested my code and ensured it meets the acceptance criteria of the story
- [x] I have tested that my code works with both the brapi-java-server and BreedBase
- [ ] I have create/modified unit tests to cover this change
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to documentation
- [ ] I have run TAF: _\<please include a link to TAF run>_

### CODE REVIEW

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java [hunk 1]
```diff
@@ -205,7 +206,8 @@ public void getExistingBrapiData(List<BrAPIImport> importRows, Program program)
 
     @Override
     public Map<String, ImportPreviewStatistics> process(List<BrAPIImport> importRows,
-                        Map<Integer, PendingImport> mappedBrAPIImport, Program program, User user, boolean commit) throws ValidatorException {
+                                                        Map<Integer, PendingImport> mappedBrAPIImport, Table table,
```
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java]: Nope, I will change them all to data to be consistent. Pushed changes.

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java [hunk 2]
```diff
@@ -145,12 +162,54 @@ public void getExistingBrapiData(List<BrAPIImport> importRows, Program program)
...
             Map<Integer, PendingImport> mappedBrAPIImport,
+            Table data,
             Program program,
             User user,
             boolean commit) throws ValidatorException, MissingRequiredInfoException {
 
         ValidationErrors validationErrors = new ValidationErrors();
 
+        // Get dynamic phenotype columns for processing
+        List<Column<?>> phenotypeCols = fileMappingUtil.getDynamicColumns(data, "ExperimentsTemplateMap");
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java:173]: Is there a constant that's defined anywhere for the `ExperimentsTemplateMap`? If not, can that be added?

File: src/main/java/org/breedinginsight/utilities/BrAPIDAOUtil.java [hunk 3]
```diff
@@ -47,12 +52,24 @@ public class BrAPIDAOUtil {
...
-                                    U searchBody
+                                                                               Function4<BrAPIWSMIMEDataTypes, String, Integer, Integer, ApiResponse<Pair<Optional<T>, Optional<BrAPIAcceptedSearchResponse>>>> searchGetMethod,
+                                                                               U searchBody
     ) throws ApiException {
+        return searchInternal(searchMethod, null, searchGetMethod, searchBody);
+    }
 
+    private <T, U extends BrAPISearchRequestParametersPaging, V> List<V> searchInternal(Function<U, ApiResponse<Pair<Optional<T>, Optional<BrAPIAcceptedSearchResponse>>>> searchMethod,
+                                                                                        Function3<String, Integer, Integer, ApiResponse<Pair<Optional<T>, Optional<BrAPIAcceptedSearchResponse>>>> searchGetMethod3,
+                                                                                        Function4<BrAPIWSMIMEDataTypes, String, Integer, Integer, ApiResponse<Pair<Optional<T>, Optional<BrAPIAcceptedSearchResponse>>>> searchGetMethod4,
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/utilities/BrAPIDAOUtil.java:71]: For naming of these to parameters, what about:
- `searchGetMethod3` goes back to `searchGetMethod`
- `searchGetMethod4` changes to `searchGetMethodWithMimeType`

**Review by timparsons — CHANGES_REQUESTED** [review]

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java (hunk 2 above)
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java]: No, I'll add one. Pushed changes.

**Review by timparsons — CHANGES_REQUESTED** [review]
I had tried importing 4 files:

- [BI_1183_valid_columns.csv](https://github.com/Breeding-Insight/bi-api/files/9748497/BI_1183_valid_columns.csv)

Result: FAIL - Error during validation:

```
java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:141)
	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.validateObservationValue(ExperimentProcessor.java:757)
	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.process(ExperimentProcessor.java:209)
	at org.breedinginsight.brapps.importer.services.processors.ProcessorManager.process(ProcessorManager.java:65)
	at org.breedinginsight.brapps.importer.model.imports.experimentObservation.ExperimentImportService.process(ExperimentImportService.java:71)
	at org.breedinginsight.brapps.importer.services.FileImportService.lambda$processFile$4(FileImportService.java:406)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1771)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1763)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

- [BI_1183_valid_with_data.csv](https://github.com/Breeding-Insight/bi-api/files/9748498/BI_1183_valid_with_data.csv)

Result: PASS - was presented with the preview table

- [BI_1183_invalid_columns.csv](https://github.com/Breeding-Insight/bi-api/files/9748499/BI_1183_invalid_columns.csv)

Result: PASS - Error showing unrecognized columns

- [BI_1183_invalid_data.csv](https://github.com/Breeding-Insight/bi-api/files/9748500/BI_1183_invalid_data.csv)

Result: FAIL - Error during validation:

```
java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:141)
	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.validateObservationValue(ExperimentProcessor.java:757)
	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.process(ExperimentProcessor.java:209)
	at org.breedinginsight.brapps.importer.services.processors.ProcessorManager.process(ProcessorManager.java:65)
	at org.breedinginsight.brapps.importer.model.imports.experimentObservation.ExperimentImportService.process(ExperimentImportService.java:71)
	at org.breedinginsight.brapps.importer.services.FileImportService.lambda$processFile$4(FileImportService.java:406)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1771)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1763)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**Comment by nickpalladino** [comment]
> I had tried importing 4 files:
> 
>     * [BI_1183_valid_columns.csv](https://github.com/Breeding-Insight/bi-api/files/9748497/BI_1183_valid_columns.csv)
> 
> 
> Result: FAIL - Error during validation:
> 
> ```
> java.util.NoSuchElementException: No value present
> 	at java.base/java.util.Optional.get(Optional.java:141)
> 	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.validateObservationValue(ExperimentProcessor.java:757)
> 	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.process(ExperimentProcessor.java:209)
> 	at org.breedinginsight.brapps.importer.services.processors.ProcessorManager.process(ProcessorManager.java:65)
> 	at org.breedinginsight.brapps.importer.model.imports.experimentObservation.ExperimentImportService.process(ExperimentImportService.java:71)
> 	at org.breedinginsight.brapps.importer.services.FileImportService.lambda$processFile$4(FileImportService.java:406)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1771)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1763)
> 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
> 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
> 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
> 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
> ```
> 
>     * [BI_1183_valid_with_data.csv](https://github.com/Breeding-Insight/bi-api/files/9748498/BI_1183_valid_with_data.csv)
> 
> 
> Result: PASS - was presented with the preview table
> 
>     * [BI_1183_invalid_columns.csv](https://github.com/Breeding-Insight/bi-api/files/9748499/BI_1183_invalid_columns.csv)
> 
> 
> Result: PASS - Error showing unrecognized columns
> 
>     * [BI_1183_invalid_data.csv](https://github.com/Breeding-Insight/bi-api/files/9748500/BI_1183_invalid_data.csv)
> 
> 
> Result: FAIL - Error during validation:
> 
> ```
> java.util.NoSuchElementException: No value present
> 	at java.base/java.util.Optional.get(Optional.java:141)
> 	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.validateObservationValue(ExperimentProcessor.java:757)
> 	at org.breedinginsight.brapps.importer.services.processors.ExperimentProcessor.process(ExperimentProcessor.java:209)
> 	at org.breedinginsight.brapps.importer.services.processors.ProcessorManager.process(ProcessorManager.java:65)
> 	at org.breedinginsight.brapps.importer.model.imports.experimentObservation.ExperimentImportService.process(ExperimentImportService.java:71)
> 	at org.breedinginsight.brapps.importer.services.FileImportService.lambda$processFile$4(FileImportService.java:406)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1771)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1763)
> 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
> 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
> 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
> 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
> ```
> 
> Traits used: [bi_ontology_template_v11_2021-12-01-test-ontology-with-method-desc.xls](https://github.com/Breeding-Insight/bi-api/files/9748517/bi_ontology_template_v11_2021-12-01-test-ontology-with-method-desc.xls)

Pushed fix for failing cases

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java [hunk 4]
```diff
@@ -708,6 +769,76 @@ private String simpleStudyName(String scopedName){
...
+        try {
+            number = new BigDecimal(value);
+        } catch (NumberFormatException e) {
+            return Optional.empty();
+        }
+        return Optional.of(number);
+    }
+
+    private boolean validNumericRange(BigDecimal value, Scale validValues) {
+        if (value.compareTo(BigDecimal.valueOf(validValues.getValidValueMin())) >= 0 &&
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java:820]: This code will throw a `NullPointerException` if `validValues.getValidValueMin` or `validValues.getValidValueMax` are `null`

**Review by timparsons — CHANGES_REQUESTED** [review]

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java [hunk 5]
```diff
@@ -708,6 +769,76 @@ private String simpleStudyName(String scopedName){
...
+                if (number.isEmpty()) {
+                    addRowError(columnHeader, "Non-numeric text detected detected", validationErrors, row);
+                }
+                else if (!validNumericRange(number.get(), variable.getScale())) {
+                    addRowError(columnHeader, "Value outside of min/max range detected", validationErrors, row);
+                }
+                break;
+            case DATE:
+                if (!validDateValue(value)) {
+                    addRowError(columnHeader, "Incorrect date format (YYYY-MM-DD) detected", validationErrors, row);
```
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java]: Sure, pushed changes.

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java (hunk 4 above)
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java]: Thanks for spotting that, pushed changes.

**Review by timparsons — APPROVED** [review]

**Pull request merged**